### PR TITLE
Consolidate E2E test suite into cli repo

### PR DESCRIPTION
**Pull request by khaong**
## Summary

- Moves the consolidated E2E test suite from `entire-cli-e2e-tests` into `e2e/` as the single E2E suite
- Removes the old shadow-hook based suite at `cmd/entire/cli/e2e_test/`
- Updates mise tasks to point at new `e2e/tests/...` path with optional filter args

## What's in `e2e/`

- **3 agents**: Claude Code, Gemini CLI, OpenCode — all run via `ForEachAgent`
- **`E2E_AGENT` env var** for CI matrix targeting (when set, only matching agent registers)
- **`//go:build e2e`** tag on all test files (invisible to normal `go build`/`go test`)
- **37 tests** covering: single/multi session, auto-commit, attribution, rewind, stash workflows, split commits, subagent flows, session lifecycle, edge cases, checkpoint metadata validation
- **Artifact capture** (git log, tree, checkpoint metadata, entire logs, console transcript) for failure debugging
- **`exploratory/`** directory for on-demand test scenarios not run by CI

## Key design decisions

- Always uses `os.MkdirTemp` instead of `t.TempDir()` — Go's nested subtest dirs confuse some agents' path resolution
- OpenCode gets `opencode.json` config written to allow `external_directory` permission in non-interactive mode
- `filepath.EvalSymlinks` resolves macOS `/var` → `/private/var` so agent CLIs see consistent paths

## Test plan

- [x] `mise run test:e2e:claude TestSingleSessionManualCommit` — PASS
- [x] `mise run test:e2e:opencode TestSingleSessionManualCommit` — PASS
- [x] `go build ./...` — clean (e2e tests invisible without build tag)
- [x] `go test -tags=e2e -list '.*' ./e2e/tests/...` — all 37 tests listed
- [x] Full suite run against Claude
- [x] Full suite run against Gemini
- [x] Full suite run agains Opencode
- [x] CI workflow update

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by khaong** [comment]
bugbot run

File: e2e/agents/claude.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Dismiss startup dialogs until we reach the input prompt.
+	for range 5 {
+		content, err := s.WaitFor(`❯`, 15*time.Second)
+		if err != nil {
+			return s, fmt.Errorf("waiting for startup prompt: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/agents/claude.go:227]: Fixed in 2f99aa85. Claude and Gemini now call `s.Close()` before returning the error, matching opencode's pattern.

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -36,8 +42,49 @@ jobs:
...
+          }
+          EOF
+
       - name: Run E2E Tests
         env:
           ANTHROPIC_API_KEY: ${{ secrets.ANTHROPIC_API_KEY }}
+          E2E_ARTIFACT_DIR: ${{ github.workspace }}/e2e-artifacts
         run: |
+          mkdir -p "$E2E_ARTIFACT_DIR"
+          entire version
```
[the remark this answers is not in the document]
Comment (khaong) [reply .github/workflows/e2e.yml]: False positive — the indentation is correct. `env:` is at step-child level (8 spaces), keys at 10 spaces, `run: |` at 8 spaces with block content at 10 spaces. This is standard GitHub Actions YAML and CI has been executing correctly with this layout.

File: e2e/agents/claude.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	envArgs := []string{"ACCESSIBLE=1", "ENTIRE_TEST_TTY=0"}
+
+	// On CI (no macOS Keychain), use an isolated config dir so Claude Code
+	// picks up ANTHROPIC_API_KEY from the environment instead of trying OAuth.
+	// Locally, we skip CLAUDE_CONFIG_DIR so the Keychain-based auth works.
+	if os.Getenv("CI") != "" {
+		configDir, err := isolatedConfigDir()
+		if err == nil {
+			envArgs = append(envArgs, "CLAUDE_CONFIG_DIR="+configDir)
+		}
```
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/agents/claude.go:205]: Acknowledged — the temp dir leaks for the lifetime of the CI runner process. Impact is negligible (a few KB in /tmp on an ephemeral runner) so not fixing for now.

File: e2e/entire/entire.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	t.Helper()
+	return runErr(dir, "rewind", "--to", id)
+}
+
+// RewindLogsOnly runs `entire rewind --to <id> --logs-only`.
+func RewindLogsOnly(t *testing.T, dir, id string) error {
+	t.Helper()
+	return runErr(dir, "rewind", "--to", id, "--logs-only")
+}
+
```
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/entire/entire.go:103]: Intentional — the `entire` package is a CLI wrapper library with functions available for future tests. Keeping `Explain`, `Resume`, etc. exported avoids reimplementing them when those test scenarios are added.

**Comment by Soph** [comment]
I tried this locally, and I couldn't get it to work at all :( 

Claude thinks: 

```
⏺ That's the bug. StartSession has the comment:
  // Locally, we skip CLAUDE_CONFIG_DIR so the Keychain-based auth works.

  But RunPrompt always sets CLAUDE_CONFIG_DIR, even locally. On macOS, Claude Code uses Keychain for auth, which is tied to the default ~/.claude path. When you override CLAUDE_CONFIG_DIR to a temp dir, Claude Code can't find auth credentials in the Keychain for that path → "Not logged in".

  RunPrompt needs the same CI-only guard that StartSession has. But this is the branch author's code, not ours. Do you want me to fix it, or flag it to the branch author?
```

I'm not sure if this is a me thing or why it works for you, so stopping trying to fix it :( 

Also: 

 1. files_touched per-checkpoint validation is gone from most tests
 2. Shadow branch cleanup checks are gone
 3. Subagent task checkpoint fields not validated (to be fair, those were logs only so fine to skip them)
 
And last but not least: it's just testing the binary in path, I feel `mise run test:e2e` makes it look like if it would test what's checked out and that's what we should do, right?

**Comment by khaong** [comment]
> I tried this locally, and I couldn't get it to work at all :(
> 
> Claude thinks:
> 
> ```
> ⏺ That's the bug. StartSession has the comment:
>   // Locally, we skip CLAUDE_CONFIG_DIR so the Keychain-based auth works.
> 
>   But RunPrompt always sets CLAUDE_CONFIG_DIR, even locally. On macOS, Claude Code uses Keychain for auth, which is tied to the default ~/.claude path. When you override CLAUDE_CONFIG_DIR to a temp dir, Claude Code can't find auth credentials in the Keychain for that path → "Not logged in".
> 
>   RunPrompt needs the same CI-only guard that StartSession has. But this is the branch author's code, not ours. Do you want me to fix it, or flag it to the branch author?
> ```
> 
> I'm not sure if this is a me thing or why it works for you, so stopping trying to fix it :(
> 
> Also:
> 
> 1. files_touched per-checkpoint validation is gone from most tests
> 2. Shadow branch cleanup checks are gone
> 3. Subagent task checkpoint fields not validated (to be fair, those were logs only so fine to skip them)
> 
> And last but not least: it's just testing the binary in path, I feel `mise run test:e2e` makes it look like if it would test what's checked out and that's what we should do, right?

@Soph - Yup, it was a real problem 🤦‍♂️ 

It turns out that I've got my claude set up just a little bit different to the 'standard OAuth' flow.

Thanks to @dvydra @toothbrush we've got to the bottom of it and fixed some other 'user scope' transgressions the tests were doing.

I think longer term we want to move these tests to docker in any case, just to be 100% certain about blast radius.

**Comment by khaong** [comment]
@Soph and as for the other points:

**1. `files_touched` per-checkpoint validation**

`TestCheckpointMetadataDeepValidation` validates `files_touched` end-to-end, and the assertion helpers (`AssertCheckpointFilesTouched`, `AssertCheckpointFilesTouchedContains`) are available for any test that needs them. Detailed per-field checkpoint validation is also covered extensively in integration tests (17 test files in `cmd/entire/cli/integration_test/`).

**2. Shadow branch cleanup checks**

`TestShadowBranchCleanedAfterAgentCommit` covers the e2e path. Detailed shadow branch lifecycle is covered in integration tests.

**Edit: I've added post-shadow checks to all the scenarios now**


**3. Testing the binary in PATH**

Intentional design - the suite is meant to run against any installed binary for version comparison and regression detection. `TestMain` prints the version and stamps `entire-version.txt` into the artifact dir. Also added the version to the mise task output so it's explicit upfront what binary is being tested.

Switched it to run against the code in the accompanying branch by default, overridable by `E2E_ENTIRE_BIN`

**Comment by khaong** [comment]
bugbot run

File: e2e/agents/tmux.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	for _, v := range unsetEnv {
+		shellCmdSb25.WriteString("env -u " + v + " ")
+	}
+	shellCmd += shellCmdSb25.String()
+	shellCmd += command
+	var shellCmdSb29 strings.Builder
+	for _, a := range args {
+		shellCmdSb29.WriteString(" " + a)
+	}
+	shellCmd += shellCmdSb29.String()
```
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/agents/tmux.go]: Fixed in c920e463. Rewrote the shell command construction to use a `parts` slice with `shellQuote()` for proper POSIX quoting. The auto-generated variable names are gone.

File: e2e/agents/claude.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/agents/claude.go:205]: Fixed in c920e463. Added `TmuxSession.OnClose()` callback mechanism and registered config dir cleanup there. Also handles the error path (cleanup if `NewTmuxSession` fails).

File: e2e/agents/opencode.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+		// Wait for TUI to be ready (input area with placeholder text).
+		if _, err := s.WaitFor(`Ask anything`, 15*time.Second); err != nil {
+			content := s.Capture()
+			_ = s.Close()
+			if strings.TrimSpace(content) == "" && attempt == 0 {
+				lastErr = err
+				continue
+			}
+			return s, fmt.Errorf("waiting for startup: %w", err)
```
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/agents/opencode.go]: Fixed in c51f1d02. Changed `return s, fmt.Errorf(...)` to `return nil, fmt.Errorf(...)` after `s.Close()` so callers don't receive an already-closed session.

File: e2e/agents/claude.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+	if errors.As(err, &exitErr) {
+		exitCode = exitErr.ExitCode()
+	}
+
+	return Output{
+		Command:  c.Binary() + " " + strings.Join(displayArgs, " "),
+		Stdout:   stdout.String(),
+		Stderr:   stderr.String(),
+		ExitCode: exitCode,
+	}, err
```
[the remark this answers is not in the document]
Comment (khaong) [reply e2e/agents/claude.go:178]: Fixed in c51f1d02. Claude and Gemini now set `ExitCode = -1` for non-ExitError failures (binary not found, context cancelled), matching OpenCode's existing behavior.

File: mise.toml [hunk 8]
```diff
@@ -1,8 +1,9 @@
 [tools]
 # Please also keep the version aligned in the go.mod file
-go = { version = '1.25.6', postinstall = "go install github.com/go-delve/delve/cmd/dlv@latest" }
+go = { version = '1.25.6', postinstall = "go install github.com/go-delve/delve/cmd/dlv@latest && go install gotest.tools/gotestsum@latest" }
```
[the remark this answers is not in the document]
Comment (khaong) [reply mise.toml:3]: Valid concern but pre-existing — the `@latest` pinning in mise.toml's `postinstall` was there before this PR. Out of scope for this change; tracked separately.

**Comment by khaong** [comment]
so enabling the shadow branch leftover checks have surfaced cases where we're not cleaning up properly - these are showing as failing tests:
`TestModifiedFileAlwaysGetsCheckpoint`
`TestContentOverlapRevertNewFile`

we are choosing to leave these in as bug markers

**Comment by Soph** [comment]
I had to run `brew install gotestsum`

**Comment by khaong** [comment]
> I had to run `brew install gotestsum`

mise install should handle this I think?

**Comment by khaong** [comment]
> TestContentOverlapRevertNewFile
> 
>   The AssertNoShadowBranches assertion is incorrect for this test and should be removed.
> 
>   This test deliberately tests a scenario where no checkpoint should be created (content mismatch on a new file). When no checkpoint is condensed, the shadow branch correctly remains
>   - it preserves the agent's discarded work for potential later access via entire rewind.

I think the mismatch here is how we're using the one-shot RunPrompt (which is 'ending' the session), I'll do a follow up where we hold the session open to better reflect the scenario. 👍
